lib/Internal/Posix/Runner.php

```diff
@@ -0,0 +1,197 @@
+            $command
+        );
+
+        $handle = new Handle;
+        $handle->proc = @\proc_open($command, self::FD_SPEC, $pipes, $cwd ?: null, $env ?: null, $options);
+
+        if (!\is_resource($handle->proc)) {
+            $message = "Could not start process";
+            if ($error = \error_get_last()) {
+                $message .= \sprintf(" Errno: %d; %s", $error["type"], $error["message"]);
+            }
+            throw new ProcessException($message);
+        }
+
+        $status = \proc_get_status($handle->proc);
+
+        if (!$status) {
+            \proc_close($handle->proc);
+            throw new ProcessException("Could not get process status");
+        }
+
+        $stdinDeferred = new Deferred;
+        $handle->stdin = new ProcessOutputStream($stdinDeferred->promise());
+
+        $stdoutDeferred = new Deferred;
+        $handle->stdout = new ProcessInputStream($stdoutDeferred->promise());
+
+        $stderrDeferred = new Deferred;
+        $handle->stderr = new ProcessInputStream($stderrDeferred->promise());
+
+        $handle->extraDataPipe = $pipes[3];
+
+        \stream_set_blocking($pipes[3], false);
+
+        $handle->extraDataPipeStartWatcher = Loop::onReadable($pipes[3], [self::class, 'onProcessStartExtraDataPipeReadable'], [$handle, [
+            new ResourceOutputStream($pipes[0]),
+            new ResourceInputStream($pipes[1]),
+            new ResourceInputStream($pipes[2]),
+        ], [
+            $stdinDeferred, $stdoutDeferred, $stderrDeferred
+        ]]);
+
+        $handle->extraDataPipeWatcher = Loop::onReadable($pipes[3], [self::class, 'onProcessEndExtraDataPipeReadable'], $handle);
+        Loop::unreference($handle->extraDataPipeWatcher);
+        Loop::disable($handle->extraDataPipeWatcher);
+
+        return $handle;
+    }
+
+    /** @inheritdoc */
+    public function join(ProcessHandle $handle): Promise {
+        /** @var Handle $handle */
+        if ($handle->extraDataPipeWatcher !== null) {
+            Loop::reference($handle->extraDataPipeWatcher);
+        }
+
+        return $handle->joinDeferred->promise();
+    }
+
+    /** @inheritdoc */
+    public function kill(ProcessHandle $handle) {
+        /** @var Handle $handle */
+        if ($handle->extraDataPipeWatcher !== null) {
+            Loop::cancel($handle->extraDataPipeWatcher);
+            $handle->extraDataPipeWatcher = null;
+        }
+
+        /** @var Handle $handle */
+        if ($handle->extraDataPipeStartWatcher !== null) {
+            Loop::cancel($handle->extraDataPipeStartWatcher);
+            $handle->extraDataPipeStartWatcher = null;
+        }
+
+        if (!\proc_terminate($handle->proc, 9)) { // Forcefully kill the process using SIGKILL.
+            throw new ProcessException("Terminating process failed");
+        }
+
+        $handle->status = ProcessStatus::ENDED;
+        $handle->joinDeferred->fail(new ProcessException("The process was killed"));
```

Comment:
I don't think that's what should happen here - we should return all the remaining data on the pipe until now instead.